Problem: missing account state initialisation and TX processing (CRO-96 CRO-97)

tomtau opened the pull request:
Solution:

* changed InitChain / genesis to start off with a global state of accounts instead of zero-input transfer transactions
* dev-utils updated to take a single json file instead of the growing number of genesis arguments
* returning updated accounts in tx validation (that are later persisted in deliver_tx / commit)

@devashishdxt I'm planning to add one more test that will go through the whole:
WithdrawUnbondedTx => TransferTx => DepositBondTx => UnbondTx on ABCI.
But since this is a big change, you can start looking at this PR now.
I'll tag master later, as this will "break" client functionality (most of the functionality should be +- OK, only the genesis processing would require some more work in client index)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/crypto-com/chain/118)
<!-- Reviewable:end -->

tomtau commented:
@devashishdxt https://reviewable.io/reviews/crypto-com/chain/118#-

tomtau commented:
@calvinaco this PR will also break the integration test, I guess -- shall we disable / allow it to fail temporarily ?

calvinaco commented:
I agree, let's temporarily change the integration tests to allow failure.

This PR involves changes to both genesis generation and `dev-utils`. There is a script to re-generate the Dockerfile and configurations when new genesis generation algorithm is changed, but it makes use of `dev-utils` which is changed in this PR as well. Once this series of change is stablized I can update the integration tests.

tomtau commented:
@devashishdxt I've added the test: https://github.com/crypto-com/chain/pull/118/commits/539b200b406fbd4d778316a7fa27fb23db8639ae found and fixed one bug through it. you can have a quick look through reviewable on the changes.
if OK, I'll tag master and squash-merge this PR

devashishdxt commented:
___
*[chain-core/src/init/coin.rs, line 154 at r3](https://reviewable.io/reviews/crypto-com/chain/118#-LgWuEZLDeQa-vRBuANH:-LgWuEZM2NN-YpvssTAe:bym3hnu) ([raw file](https://github.com/crypto-com/chain/blob/4467ec14f34d2caf831c320924b18d034204e8f2/chain-core/src/init/coin.rs#L154)):*
> ```Rust
> }
> 
> impl From<Coin> for i64 {
> ```

Why not implement `TryFrom` instead of `From`?


<!-- Sent from Reviewable.io -->

devashishdxt commented:
___
*[dev-utils/src/commands/genesis_command.rs, line 121 at r1](https://reviewable.io/reviews/crypto-com/chain/118#-LgWsVELBhVLk0EhBtre:-LgWsVELBhVLk0EhBtrf:b-tmcdok) ([raw file](https://github.com/crypto-com/chain/blob/047bee7ead88ce1838b4057577f06917085ff16c/dev-utils/src/commands/genesis_command.rs#L121)):*
> ```Rust
>             println!("\"app_state\": {}", config_str);
>         } else {
>             panic!("distribution validation error: {}", result.err().unwrap());
> ```

Return `Error` here?


<!-- Sent from Reviewable.io -->

devashishdxt commented:
Is `chain-core/tests/distribution.json` necessary to commit? If yes, can we reduce its size?

tomtau commented:
> Is `chain-core/tests/distribution.json` necessary to commit? If yes, can we reduce its size?
it's not necessary -- just tried real-world data for testing.
I could try to trim it.
